esphome/components/wts01/wts01.cpp

```diff
@@ -0,0 +1,100 @@
+#include "wts01.h"
+#include "esphome/core/log.h"
+#include <cmath>
+
+namespace esphome {
+namespace wts01 {
+
+static const char *const TAG = "wts01";
+static const uint8_t PACKET_SIZE = 9;
+static const uint8_t HEADER_1 = 0x55;
+static const uint8_t HEADER_2 = 0x01;
+static const uint8_t HEADER_3 = 0x01;
+static const uint8_t HEADER_4 = 0x04;
+
+void WTS01Sensor::loop() {
+  // Process one character at a time received from the sensor
+  if (available()) {
+    uint8_t c;
+    if (read_byte(&c)) {
+      this->handle_char_(c);
+    }
+  }
+}
+
+void WTS01Sensor::dump_config() {
+  ESP_LOGCONFIG(TAG, "WTS01 Temperature Sensor:");
+  LOG_SENSOR("  ", "Temperature", this);
+  this->check_uart_settings(9600);
+}
+
+void WTS01Sensor::handle_char_(uint8_t c) {
+  // Reset buffer if we're at the start of a new packet
+  if (c == HEADER_1 && this->buffer_pos_ == 0) {
+    this->buffer_[this->buffer_pos_++] = c;
+    return;
+  }
+
+  // Check header sequence
+  if (this->buffer_pos_ == 1 && c != HEADER_2) {
+    this->buffer_pos_ = 0;
+    return;
+  }
+
+  if (this->buffer_pos_ == 2 && c != HEADER_3) {
+    this->buffer_pos_ = 0;
+    return;
+  }
+
+  if (this->buffer_pos_ == 3 && c != HEADER_4) {
+    this->buffer_pos_ = 0;
+    return;
+  }
+
+  // Add byte to buffer
+  this->buffer_[this->buffer_pos_++] = c;
+
+  // Process complete packet
+  if (this->buffer_pos_ == PACKET_SIZE) {
```

Comment:
Better to have a greater-or-equal-check when dealing with a buffer position counter like this. This change will reduce the chance of accidentally overrunning the buffer in case of any logic errors.
```suggestion
  if (this->buffer_pos_ >= PACKET_SIZE) {
```